godot-ffi/src/godot_ffi.rs

```diff
@@ -19,74 +19,56 @@ use std::marker::PhantomData;
     }
 
     /// Construct from Godot opaque pointer.
     ///
+    /// This will increment reference counts if the type is reference counted. If you need to avoid this then a `borrow_sys` method should
+    /// usually be used. Which is a method that takes a sys-pointer and returns it as a `&Self` reference. This must be manually implemented for
+    /// each relevant type as not all types can be borrowed like this.
+    ///
     /// # Safety
     /// `ptr` must be a valid _type ptr_: it must follow Godot's convention to encode `Self`,
     /// which is different depending on the type.
-    /// The type in `ptr` must not require any special consideration upon referencing. Such as
-    /// incrementing a refcount.
-    unsafe fn from_sys(ptr: sys::GDExtensionTypePtr) -> Self;
+    #[doc(hidden)]
+    unsafe fn new_from_sys(ptr: sys::GDExtensionConstTypePtr) -> Self;
 
     /// Construct uninitialized opaque data, then initialize it with `init_fn` function.
     ///
     /// # Safety
     /// `init_fn` must be a function that correctly handles a (possibly-uninitialized) _type ptr_.
-    unsafe fn from_sys_init(init_fn: impl FnOnce(sys::GDExtensionUninitializedTypePtr)) -> Self;
+    #[doc(hidden)]
+    unsafe fn new_with_uninit(init_fn: impl FnOnce(sys::GDExtensionUninitializedTypePtr)) -> Self;
 
-    /// Like [`Self::from_sys_init`], but pre-initializes the sys pointer to a `Default::default()` instance
-    /// before calling `init_fn`.
+    /// Like [`new_with_uninit`](GodotFfi::new_with_uninit), but pre-initializes the sys pointer to a default instance (usually
+    /// [`Default::default()`]) before calling `init_fn`.
     ///
     /// Some FFI functions in Godot expect a pre-existing instance at the destination pointer, e.g. CoW/ref-counted
     /// builtin types like `Array`, `Dictionary`, `String`, `StringName`.
     ///
-    /// If not overridden, this just calls [`Self::from_sys_init`].
-    ///
     /// # Safety
-    /// `init_fn` must be a function that correctly handles a (possibly-uninitialized) _type ptr_.
-    unsafe fn from_sys_init_default(init_fn: impl FnOnce(sys::GDExtensionTypePtr)) -> Self
-    where
-        Self: Sized, // + Default
-    {
-        // SAFETY: this default implementation is potentially incorrect.
-        // By implementing the GodotFfi trait, you acknowledge that these may need to be overridden.
-        Self::from_sys_init(|ptr| init_fn(sys::AsUninit::force_init(ptr)))
-
-        // TODO consider using this, if all the implementors support it
-        // let mut result = Self::default();
-        // init_fn(result.sys_mut().as_uninit());
-        // result
-    }
+    /// While this does call `init_fn` with a `&mut Self` reference, that value may have a broken safety invariant.
+    /// So `init_fn` must ensure that the reference passed to it is fully initialized when the function returns.
+    #[doc(hidden)]
+    unsafe fn new_with_init(init_fn: impl FnOnce(&mut Self)) -> Self;
```

Comment:
Is there an example where such an invariant is necessary?

So it's generally not possible to make this method safe?

Reply:
I believe `RawGd` requires it, atm with how its safety invariant is structured. Since for `RawGd` the rtti info and object pointer must remain unchanged after construction, but if you use this method you'd need to overwrite them with more appropriate values. i can check if it's possible to rework some things to make it safe though.

Reply:
Ah don't worry, if there's a concrete case that needs it, it's fine to keep `unsafe`. Maybe add a brief comment, I was just not sure if this was leftover from refactoring.

Reply:
Actually thinking about this, does `RawGd` even need it? like sure we set the object pointer and rtti info to be a null-object. But as long as we're careful to never let this constructed `RawGd` leak to user-code before we change the object pointer/rtti info to the correct info, then it is impossible to tell that the `RawGd` had this happen to it.

So i guess yeah it can be a safe function, it just means that user-code could only ever use it to construct a null-object. Which is probably fine?